source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2186,26 +2185,52 @@ count = (first_count.resname.count - std::max(0, static_cast<int32_t>(first_coun
                 CLAMP_COUNT(textureImages, COpenGLExtensionHandler::maxImageBindings, image);
                 newImgCount = first_count.textureImages.first + count;
 #undef CLAMP_COUNT
+            }
 
+            if (i < compatibilityLimitPlusOne)//if prev and curr pipeline layouts are compatible for set N and currState.set[N]==nextState.set[N] then binding set N would be redundant
                 continue;
-            }
 
             const auto& multibind_params = nextState.descriptorsParams.descSets[i].set ? 
                 nextState.descriptorsParams.descSets[i].set->getMultibindParams() :
                 COpenGLDescriptorSet::SMultibindParams{};//all nullptr
 
-            //TODO dynamic offsets
-
-            extGlBindBuffersRange(GL_UNIFORM_BUFFER, first_count.ubos.first, (newUboCount - first_count.ubos.first), multibind_params.ubos.buffers, multibind_params.ubos.offsets, multibind_params.ubos.sizes);
-            extGlBindBuffersRange(GL_SHADER_STORAGE_BUFFER, first_count.ssbos.first, (newSsboCount - first_count.ssbos.first), multibind_params.ubos.buffers, multibind_params.ubos.offsets, multibind_params.ubos.sizes);
+            const GLsizei localUboCount = (newUboCount - first_count.ubos.first);//"local" as in this DS
+            const GLsizei localSsboCount = (newSsboCount - first_count.ssbos.first);//"local" as in this DS
+            //not entirely sure those MAXes are right
```

Comment:
not quite (they are combined maxes across all pipeline stages) but they aren't obscene.